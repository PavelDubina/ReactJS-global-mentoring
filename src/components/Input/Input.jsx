import React from 'react'
import PropTypes from 'prop-types'
import styles from './Input.scss'

export const Input = ({ title, name, type, value = '', onChange, placeholder = '', style = {}, disabled = false }) => (
  <label className={styles.label}>
    {title}
    <input
      className={styles.input}
      placeholder={placeholder}
      name={name}
      type={type}
      value={value}
      disabled={disabled}
      style={style}
      onChange={onChange}
      autoComplete="off"
    />
  </label>
)

Input.propTypes = {
  title: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  placeholder: PropTypes.string,
  style: PropTypes.objectOf(PropTypes.string),
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
}
