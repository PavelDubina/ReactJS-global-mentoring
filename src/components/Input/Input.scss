.label {
  display: flex;
  flex-direction: column;
}

.input {
  margin: 8px 0 0 0;
  padding: 0 16px;
  background-color: #232323c9;
  border-radius: 5px;
  border-width: 0;
  font-size: 24px;
  line-height: 48px;
  color: #fff;
  outline: none;
}

input[type='number'] {
  padding: 0;
  margin: 0;
  background-color: #141414;
}

input[type='date']::-webkit-calendar-picker-indicator {
  cursor: pointer;
  opacity: 0.6;
  outline: none;
  filter: invert(1);

  &:hover {
    opacity: 1;
  }
}
