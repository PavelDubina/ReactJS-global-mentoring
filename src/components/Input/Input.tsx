import React from 'react'
import { useField } from 'formik'
import styles from '../../../styles/Input.module.scss'

type InputProps = {
  title: string
  name: string
  type: string
  value: string | number
  placeholder?: string
  style?: { [key: string]: string }
  disabled?: boolean
}

export const Input: React.FC<InputProps> = ({ title, ...props }) => {
  const [field, meta] = useField(props)
  return (
    <label className={styles.label}>
      {title}
      <input data-testid={props.name} {...field} {...props} className={styles.input} autoComplete="off" />
      {meta.touched && meta.error ? (
        <p data-testid="error" className={styles.error}>
          {meta.error}
        </p>
      ) : null}
    </label>
  )
}
