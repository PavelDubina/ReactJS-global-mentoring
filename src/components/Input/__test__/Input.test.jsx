import React from 'react'
import { render } from '@testing-library/react'
import { useField } from 'formik'
import { Input } from '../Input'

jest.mock('formik')

describe('Input', () => {
  const mockMeta = {
    touched: false,
    error: '',
  }
  const mockField = {
    onChange: jest.fn(),
  }
  useField.mockReturnValue([mockField, mockMeta])

  const getComponent = ({
    title = 'test title',
    name = 'test name',
    type = 'text',
    value = 'test value',
    placeholder = 'test placeholder',
  }) => render(<Input title={title} name={name} type={type} value={value} placeholder={placeholder} />)
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('Should show error', () => {
    const mockMeta = {
      touched: true,
      error: 'This is test error',
    }
    const mockField = {
      onChange: jest.fn(),
    }
    useField.mockReturnValue([mockField, mockMeta])
    const { getByText } = getComponent({})
    expect(getByText('This is test error')).toBeTruthy()
  })
})
