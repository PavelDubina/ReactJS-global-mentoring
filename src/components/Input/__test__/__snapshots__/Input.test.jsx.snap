// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Should renders correctly 1`] = `
<DocumentFragment>
  <label
    class="label"
  >
    test title
    <input
      autocomplete="off"
      class="input"
      data-testid="test name"
      name="test name"
      placeholder="test placeholder"
      type="text"
      value="test value"
    />
  </label>
</DocumentFragment>
`;
